src/workerd/api/streams/writable.c++

```diff
@@ -369,6 +369,171 @@ private:
+        }
+      }
+    }));
+  }
+
+  kj::Promise<void> write(const void* buffer, size_t size) override {
+    auto pieces = kj::arr(kj::arrayPtr(static_cast<const kj::byte*>(buffer), size));
+    return write(pieces).attach(kj::mv(pieces));
+  }
+
+  kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const byte>> pieces) override {
+    if (writer == kj::none) {
+      return KJ_EXCEPTION(FAILED, "Write after stream has been closed.");
+    }
+    auto amount = 0;
+    for (auto& piece : pieces) {
+      amount += piece.size();
+    }
+    if (amount == 0) return kj::READY_NOW;
+    return canceler.wrap(context.run([this, amount, pieces](Worker::Lock& lock) mutable {
+      auto& writer = getInner();
+      // Sadly, we have to allocate and copy here. Our received set of buffers are only
+      // guaranteed to live until the returned promise is resolved, but the application code
+      // may hold onto the ArrayBuffer for longer. We need to make sure that the backing store
+      // for the ArrayBuffer remains valid.
+      auto source = KJ_ASSERT_NONNULL(jsg::BufferSource::tryAlloc(lock, amount));
+      auto ptr = source.asArrayPtr();
+      for (auto& piece : pieces) {
+        KJ_DASSERT(ptr.size() > 0);
+        KJ_DASSERT(piece.size() <= ptr.size());
+        if (piece.size() == 0) continue;
+        ptr.slice(0, piece.size()).copyFrom(piece);
+        ptr = ptr.slice(piece.size());
+      }
+
+      return context.awaitJs(lock, writer.write(lock, source.getHandle(lock)));
+    }));
+  }
+
+  // TODO(perf): We can't properly implement tryPumpFrom(), which means that Cap'n Proto will
+  //   be unable to perform path shortening if the underlying stream turns out to be another capnp
+  //   stream. This isn't a huge deal, but might be nice to enable someday. It may require
+  //   significant refactoring of streams.
+
+  kj::Promise<void> whenWriteDisconnected() override {
+    // TODO(soon): We might be able to support this by following the writer.closed promise,
+    // which becomes resolved when the writer is used to close the stream, or rejects when
+    // the stream has errored. However, currently, we don't have an easy way to do this.
+    //
+    // The Writer's getClosed() method returns a jsg::MemoizedIdentity<jsg::Promise<void>>.
+    // jsg::MemoizedIdentity lazily converts the jsg::Promise into a v8::Promise once it
+    // passes through the type wrapper. It does not give us any way to consistently get
+    // at the underlying jsg::Promise<void> or the mapped v8::Promise. We would need to
+    // capture a TypeHandler in here and convert each time to one or the other, then
+    // attach our continuation. It's doable but a bit of a pain.
+    //
+    // For now, let's handle this the same as WritableStreamRpcAdapter and just return a
+    // never done.
+    return kj::NEVER_DONE;
+  }
+
+  kj::Promise<void> end() override {
+    ended = true;
+    if (writer == kj::none) return kj::READY_NOW;
```

Comment:
Will make it throw. I think I prefer just checking to see if the writer is `kj::none` here directly and throwing a more specific error message rather than the one thrown by `getInner()` but the effect will be the same.